src/lib.rs

```diff
@@ -12,61 +14,130 @@ pub struct LinkDrop {
     pub accounts: Map<PublicKey, Balance>,
 }
 
-const ACCESS_KEY_ALLOWANCE: u128 = 100_000_000_000_000_000_000;
+/// Access key allowance for linkdrop keys.
+const ACCESS_KEY_ALLOWANCE: u128 = 1_000_000_000_000_000_000_000;
+
+/// Gas attached to the callback from account creation.
+pub const ON_CREATE_ACCOUNT_CALLBACK_GAS: u64 = 100_000_000_000_000;
+
+/// Indicates there are no deposit for a callback for better readability.
+const NO_DEPOSIT: u128 = 0;
 
-fn str_to_public_key(public_key: String) -> PublicKey {
-    let mut pk = vec![0];
-    pk.extend(bs58::decode(public_key).into_vec().unwrap());
-    pk
+#[ext_contract(ext_self)]
+pub trait ExtLinkDrop {
+    fn on_account_created_and_claimed(&mut self) -> bool;
+}
+
+fn is_promise_success() -> bool {
+    assert_eq!(
+        env::promise_results_count(),
+        1,
+        "Contract expected a result on the callback"
+    );
+    match env::promise_result(0) {
+        PromiseResult::Successful(_) => true,
+        _ => false,
+    }
 }
 
 #[near_bindgen]
 impl LinkDrop {
     /// Allows given public key to claim sent balance.
-    pub fn send(&mut self, public_key: String) -> Promise {
-        let pk = str_to_public_key(public_key);
-        self.accounts.insert(&pk, &env::attached_deposit());
-        Promise::new(env::current_account_id())
-            .add_access_key(
-                pk,
-                ACCESS_KEY_ALLOWANCE,
-                env::current_account_id(),
-                "claim,create_account_and_claim,promise_batch_action_add_key_with_function_call"
-                    .to_string()
-                    .into_bytes(),
-            )
+    #[payable]
+    pub fn send(&mut self, public_key: Base58PublicKey) -> Promise {
+        let pk = public_key.into();
+        let value = self.accounts.get(&pk).unwrap_or(0);
+        self.accounts
+            .insert(&pk, &(value + env::attached_deposit()));
+        Promise::new(env::current_account_id()).add_access_key(
+            pk,
+            ACCESS_KEY_ALLOWANCE,
+            env::current_account_id(),
+            "claim,create_account_and_claim".to_string().into_bytes(),
+        )
     }
 
-    /// Claim tokens that are attached to the public key this tx is signed with.
-    pub fn claim(&mut self) -> Promise {
+    /// Claim tokens for specific account that are attached to the public key this tx is signed with.
+    pub fn claim(&mut self, account_id: AccountId) -> Promise {
```

Comment:
Check the validity of the `account_id`

Reply:
Should we add the check into AccountId parsing?

Reply:
I thought about this, but decided it's not the right place, but actually you want to check the validity most of the time. So we should introduce `ValidAccountId` type